### ISSUE TITLE
Opaque / Non opaque access token

### ISSUE THREAD

**Issue by fimbault** [ietf-wg-gnap/gnap-core-protocol#145]
Currently we assume that tokens are opaque to the client, meaning that the client doesn't care about their internals (although in practice JWT is often used). It's even described as such in our current version of the terminology (v02) : "The contents and format of the access token are opaque to the RC."

Yet, there has been discussion on the IETF mailing list that for privacy reasons, there might be a need for non opaque tokens, so that its content can be verified (i.e. doesn't include more rights than requested). 

Related actions : do we want to support non opaque tokens?  
1) review the potential impact on the definition of "access token" in #29 
2) should we define a datastructure for GNAP access tokens? (JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens or equivalent). How do we keep flexibility if one needs to extend the default format (more fields), or even integrate with something different than JWT? (we don't want to limit potential innovation here).
3) does it replace introspection/management endpoints?
4) should it be enforced? is it just an option?

**Comment by jricher** [comment]
I am against the client having any expectation of being able to understand the access token. If a component needs to send a message directly to the client in the way the client can understand it, then it should use a different protocol element to do so.

The common trust model in the OAuth world is that the client is the untrusted component, which is why it needs to get the access token in the first place.

**Comment by fimbault** [comment]
I have mixed feelings here. 
I agree with the client being unstrusted from a security perspective, what we're saying here is that the AS is untrusted from a privacy perspective. 

To me, the biggest potential downside of non opaque tokens is that it potentially limits innovation. A big part of the success of JWT comes from json's flexibility. Plus nothing stops me in OAuth from using a completely different format if I wish. That's really valuable. My current experience here would be to keep that behavior.
But it does come at the expense of not having a schema that can be verified by default, which could come handy in some cases. The alternative is introspection/management APIs, but there's always the possibility that the token could actually be different than the related resource description.  

So to go forward, I guess Denis should make its case here, and then we can take a decision. This should also balance the alternatives.

**Comment by jricher** [comment]
It's also a privacy issue to release information to the client in many cases. Several profiles of OAuth that I've worked on prohibit the token itself to carry user information or identifiers in a way that the client could read. In these cases, it's the business of the RS and the AS, and the token is just the artifact to carry rights between them. 

For example: in a medical space, the AS knows the user's account and the medical record number associated with that account. The RS just needs to know the medical record number. The client doesn't need to know either of those if it's given an appropriately scoped API and access token. None of that is possible if the token is not opaque, even optionally.

I think it might be better to approach these privacy and rights concerns by clarifying things like what gets returned alongside the access token itself, instead of inside the access token. (see #141 for one slice of this) This fits the model I mention above of sending information alongside the access token specifically for the client to consume, instead of assuming the client can look into or otherwise interpret and understand the access token itself.

**Comment by fimbault** [comment]
That's true indeed

**Comment by aaronpk** [comment]
Editor's hat off, I am 100% against the idea of letting/requiring clients interpret the contents of access tokens. Justin did a great job of illustrating examples and reasons of why.

**Comment by yaronf** [comment]
Can we allow non-opaque tokens as a (future) extension? The client asks for one specifically, and the AS (if it wants) provides a token that's guaranteed to be human-readable.

**Comment by fimbault** [comment]
This has appeared again in the terminology PR #155 

Proposal for access token : "a data artifact representing a set of delegated access rights, the format of which is opaque to the client software."
The last part of the sentence is a requirement, so should be included in the main text, not in the definition.

**Comment by Denisthemalice** [comment]
Access tokens should not considered to be opaque. Let us also suppose that an end-user wishes to use an identifier specific to a RS because he does not want RSs to link accesses on different RSs. Let us suppose that an organization implements an AS that conforms to GNAP. If access tokens were considered to be opaque, that organization could insert, in addition to the locally unique identifier, a globally unique identifier inside the access token and clients would not be allowed to detect it. So it is a matter of trust towards the AS by the client to behave correctly. Clients need to be able to control the content of the access token to make sure it only contains what has been requested and no more. If the content is not satisfactory, then the client will discard the access token. It is also interesting for the client to control that the access token is well-formed before being communicated to the RS. 

It would be wise to introduce a version number inside the access token to identify its version and allow changes in its structure for the future.

**Comment by fimbault** [comment]
@Denisthemalice in theory the risk you mention exists. But there have been several arguments against the idea. It seems unlikely that we'd move towards implementing non-opaque tokens as a general scheme. The question really is whether we have valid reasons to support it as an optional/extension feature, as suggested by Yaron. If so, we'll probably need some text to get a concrete proposal.

**Comment by Denisthemalice** [comment]
As far as I understand, the protocol between the client and the AS returns, for the access token, a data structure and not a blob. If it were a blob, we would still need some kind of version identifier so that the RS can interpret the content of the blob. Originally, in OAuth, a RS was supposed to have a strong relationship with one AS. In GNAP, I believe that an AS can deliver an access token to a RS with which he has no prior relationship. So there is not necessarily a prior agreement about the structure of the access token between the RS and an AS. The AS may publish the version(s) of the access token it supports. In the same way, the RS may publish the version(s) of the access token it supports. Then the client would  be able to select a version in the intersection of these versions (if any).

Allowing the client to inspect the access token is a matter of trust in the system. Transparency is an important privacy principle.
Without transparency, there is no trust by the clients and hence by the end-users.

So it is not a matter of risk as you say, but a matter of trust or confidence.

**Comment by fimbault** [comment]
Yes I agree it's linked to the trust model. Currently it is generally assumed that the AS is trusted, although we might indeed want to relax some of that within GNAP in the future.

**Comment by Denisthemalice** [comment]
It is not a matter of "relaxing trust". Trust is not black and white.
 
It is important to recall what means trust: degree to which a user or other stakeholder has confidence that a product 
or **system will behave as intended** 
 See ISO/IEC 25010:2011(en), 4.1.3.2

It is of particular importance that users (and auditors) can have confidence that **the AS behaves as intended**. 
Without access token transparency, they can't.

**Comment by fimbault** [comment]
Ok we all agree on the fact that there should be assurance as to how the system behaves. I'm not totally sure the proposed solution is absolutely necessary for that (there are other ways to build confidence, such as product certification for instance). 
If you want to put that issue forward, the proposed way would be to submit text for an extension and gather consensus from the WG (knowing that current cases are based on opaque tokens, which are here to stay and provide some benefits).

**Comment by Denisthemalice** [comment]
Product certification provides confidence at the time of the audit, but not later on
Token transparency provides confidence in real time to any client. This makes a big difference. 

For the time being an AS is a server. Until a secure element is used as an AS in a phone, that case does not exist. 

It is not a matter of "an extension". Nothing would prevent a client to ignore the content of an access token if it wishes to do so.
But those clients that wish to inspect the content of the access token should be prevented to do it.

Both philosophies can co-exist, without using any "extension".

**Comment by fimbault** [comment]
what I mean by extension is relative to GNAP core, which is probably going to be based on opaque tokens, at least until we can review what the alternative implies.

**Comment by Denisthemalice** [comment]
Let us explore some implications.

As indicated above, this topic is related whether or not an access token can be issued to a RS with which an AS has no prior relationship. 

With opaque tokens, a prior relationship is mandatory. This is adequate for closed environments only.
With non-opaque tokens, there is no need for a prior relationship. This allows scalability over the whole Internet.

There are also some side effects:

In the former case, the AS will be able to know and spy where the token it issues is likely to be delivered.
In the later case, it is even possible in some cases to hide to the AS the identity of the RS.

Let us wait and see what other people which are not in the same time zone think about these arguments.

**Comment by fimbault** [comment]
Handling opaque tokens would require additional text, not sure how that would work.  
I propose that we put in the text a requirement that tokens must be non opaque tokens (what is currently known to be working well), to be reviewed later on if as @yaronf suggested we find a reason and a way to cover both cases. 

See also issue #169 which advocates for the possibility of various token mechanisms, which is much easier to do if tokens are considered opaque. Would it mean we need to specify their type somewhere?

**Comment by jricher** [comment]
@fimbault I think you have that backwards -- current text is that all tokens are assumed to be opaque and handling non-opaque tokens would require additional text.

**Comment by fimbault** [comment]
I need to check but I remember I removed a definition where it was explicit. Maybe it's already somewhere else already, in that case indeed no change is needed.

**Comment by jricher** [comment]
No we pulled it out of the definition of tokens that they're opaque to the client software, and that needs to be added back in. But what's "known to be working well" is opaque tokens, and I was saying your comment above was inverted from what you meant, I think. :)

**Comment by fimbault** [comment]
yes sorry :-)

**Comment by yaronf** [comment]
This thread is getting too long... I think the Terminology section is not the right place to mandate technical properties. I thought we already found a more appropriate section to include this text.

Re: the actual question, @Denisthemalice explained why he needs non-opaque tokens, hence my question whether those can be a future extension.

**Comment by agropper** [comment]
Is it not the case that attenuated delegation requires non-opaque tokens? See also #169.

**Comment by fimbault** [comment]
As Justin pointed out rightly, I wrote my comment too fast, sorry. And so it wasn't clear, especially for delegated tokens which are a bit special. 

Delegated tokens are opaque to the protocol in general, you just can pass them around. But they wouldn't be opaque to the client that wants to execute attenuation.
So to clarify what I meant : I think opaque tokens, which wouldn't make it mandatory for the client to understand the format unless there's a very good reason for them to do so, is a better assumption. But there might still be the need to know which token we're dealing with, just in case there's a specific need for it. 

Hopefully this is a bit more clear.

**Comment by Denisthemalice** [comment]
The new current definition of an access token is now:

> Access Token 
> a data artifact representing a set of rights and/or attributes.
> Note: an access token can be first issued to an client instance (requiring authorization by the RO) and subsequently rotated.

The Note should be removed since in some cases, an authorization by the RO is necessary (i.e. when rights are being used) while in some other cases, it is unnecessary (i.e. when attributes are being used). Another argument is that this Note does not explain a facet of the definition but only one of the many usages of the access token.  

I agree with fimbault : we wouldn't make it mandatory for the clients to understand the detailed format unless there's a very good reason for them to do so, Such a good reason is confidence if the system (e.g. inspection of the content of the access token).

However, when an AS supports more than one access token format, a client should be able to request an access token compliant with a given format supported by the RS.  In any case, access tokens formats should be advertised by the AS, so that the client can verify whether there is a match with the access token formats supported by the RS.  

The AS Discovery function which is supported by sending an HTTP OPTIONS request to the grant request endpoint from an AS to retrieve the server's discovery information should be able to return the supported access token formats.

The same should apply for a RS Discovery function.

**Comment by fimbault** [comment]
The note refers to the interaction with the RO. This happens before any token is issued. 

Supposing we include that functionality:
the token format could just be a parameter in the request (access_token_types), discovery could return the access_token_types_supported.

**Comment by agropper** [comment]
Please confirm compatibility with a capabilities model. 

- The RO client may get one or more capabilities from the RS anonymously (pays in cash)
- Some of the capabilities could be subject to filtering to reduce privileges.
- Some capabilities could be directed to an auditor that would have to participate in the access transaction.
- The RO may pass one or more capabilities along with some private policies to a GNAP AS. 
- The RO may pass one or more capabilities to an auditor role.
- The AS consumes and processes requests based on RO policy to issue (filtered capabilities) as access tokens.
- The RO does not have to be accessible when the GNAP AS processes some requests (based on policy, of course).
- The access tokens are capable of further filtered delegation outside the control of AS or RS.
- When finally processed at the RS, an audit trail can be constructed as promised in the original RO / RS contract.

Adrian

**Comment by fimbault** [comment]
I don't think you can get all of that for all types of tokens (hence the need for multiple formats / depending on the needs). 
Ex: 
- JWT can't be filtered once issued
- but auditing ocaps is more complex (see for instance [zcap handling abuse](https://w3c-ccg.github.io/zcap-ld/#handling-abuse))

**Comment by agropper** [comment]
It seems obvious that the RS holds all the power in every instance. Some of
that power is shared with the RO in a GNAP-compatible way. Capabilities
need not be the only GNAP-compatible option but I hope we can have an
example of how at least one capability model is used in an audit-capable
way.

ROs need audit-ability in order to hold delegates, including the RS and AS,
to account.

Adrian

**Comment by fimbault** [comment]
It depends what you want to audit:
- For pure ocaps, auditing policies is fine but auditing access is complex. See zlcap and [Horton](http://www.erights.org/elib/capability/horton/).
- for ACL, RBAC, ABAC, implementing access control is straighforward. It's also possible to audit dynamic policies (ex; NIST NCAG, OPA)

**Comment by agropper** [comment]
I’m primarily looking to make GNAP’s role in zero-trust architecture (ZTA)
as obvious as practical.

I use the term policies exclusively as those things the RO controls that it
is willing to share with the GNAP AS and nobody else.

I look for auditability to be an option that introduces another party, for
example a notary and a log they create, that can be used to document the
outcome of a GNAP-protected access.

The audit log could document what client requested access at the AS or
received access to the RS. I do not consider this an audit of policies
themselves but rather the outcome of applying the policies.

Auditability is very different from audit. Audit can be very costly and
involve disclosure of policies and contracts that have nothing to do with
GNAP.

In ZTA terms, I’m hoping for a way to keep both the RS and the AS honest by
introducing the option for a transaction log kept by the notary.

Adrian

**Comment by fimbault** [comment]
Not sure, but my 2 cents.

We discussed in the past the idea of a separate Interact Server (IS), which is a functional role in the AS (but that could be separated physically) dedicated to managing the interactions (so, what is requested by the client). You might log those decisions. 

If that IS is managed by the RS (or a gateway to the RS), then it could potentially extend its log facility to what client really request, but that becomes a specific deployment scenario.

**Comment by agropper** [comment]
A compromised RS or AS can’t be trusted to keep accurate logs but we need
to do the best we can. On the RS side, we can encourage contemporaneous
write-only access to an external log. On the AS side, we can ensure that an
honest RS logs the action of the AS as much as the RO is willing to have be
transparent for audit purposes.

I believe this is the most common case. Two parties to a contract, the RO
and RS as client and service provider respectively, choose to introduce a
notary as a third party that keeps a log.

Adrian

**Comment by jricher** [comment]
Access tokens will continue to be opaque to the client instance. Both structured/formatted tokens an unstructured tokens need to be supported by the core protocol.

The format and data model of the access token must be known to AS, and the access token must be able to be interpreted by the RS. However, none of these requirements lead to the client instance needing to know what is inside the token itself and there are significant privacy and security issues with the client instance needing to know the structure and content of the token. Interoperability between the AS and RS in terms of token formats could be addressed in a separate extension document, akin to UMA2's Federated Authorization https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-federated-authz-2.0.html

See #114 for separating the RS-facing protocol from the core.

**Comment by agropper** [comment]
@jricher Are you saying that GNAP clients cannot support filtering capabilities they receive from an AS?

**Comment by Denisthemalice** [comment]
A typo in the following sentence:

> An end-user should be able to consent about the privileges  hat are being included into an access token and then after should be able to verify that these privileges and NO MORE have effectively been placed into the **_AS_**.


I was meaning:

> An end-user should be able to consent about the privileges that are being included into an access token and then after should be able to verify that these privileges and NO MORE have effectively been placed into the **_access token_** .

**Comment by jricher** [comment]
@Denisthemalice Since the AS creates the token, it is fully and finally responsible for what goes "into" the access token. The client can only request, the AS declares the result.

**Comment by Denisthemalice** [comment]
@jricher;

If you want to have a design based on "**_Spy by design_**" then you are right, since the end-user will not know whether, for example, his email address has been included or not into the access token. If it is included, it allows the RSs to link their user accounts and other servers that use that email address to also link their user accounts with the RSs that receive that email address. 

If you want to have a design that takes care of the end-user privacy, then you have no choice: 
**_the access token MUST NOT be opaque to the client._**

**Comment by jricher** [comment]
@Denisthemalice the access token could always be used by the RS to look up information about the user, which can include private information, so making the token non-opaque doesn't help. Your proposal also eliminates the ability to have by-reference access tokens that contain :no: information at all. This is the most used pattern in privacy preserving systems today. Furthermore, your proposal would leak information to the client instance. As a user, I don't want my client to know things like my email address or account information if it doesn't have to. 

What you're describing is best situated as privacy considerations for the AS implementation, and as part of the design of an RS-facing portion of the protocol space as discussed in #114. None of these require the access token to be non-opaque to the client, and in fact there are many documented disadvantages to doing so.

**Comment by Denisthemalice** [comment]
@jricher . You wrote:

> the access token could always be used by the RS to look up information about the user, which can include private information, so making the token non-opaque doesn't help.

As soon as an RS would be doing a Token Introspection call to the AS, the client will be ignorant whether the AS includes or not in its response, by example, an email address. A Token Introspection call to the AS made by the RS would be another component of a design based on **_"Spy by design"_** .

You also wrote: 

> Furthermore, your proposal would leak information to the client instance.

If the end-user does not trust his client, it should not use it. The end-user trusts its client to behave as expected.  I already mentioned that the draft should describe upfront the trust relationships and such a section is still currently missing.  
See "Trust relationships" Issue #214

You also wrote:

> What you're describing is best situated as privacy considerations for the AS implementation, 

This has little to do with privacy considerations for the AS implementation. The AS should be prevented, as far as possible, to act as "**_Big Brother_**". This has to do with privacy considerations for the **_end-user_** (and hence client) and for the **_RS_**. 

- The **_end-user_** (and hence the client) should be able to know which private attributes are being communicated by the AS to the RS, and the most important is that the end-user should be able to consent to the communication of these private attributes and also should be informed of the reasons to communicate them. 
- The **_RS_** should not implement nor use Token Introspection. In general, any call back from a RS to an AS should be avoided.

As I already said: 

> Transparency is an important privacy principle. Without transparency, there is no trust by the clients and hence by the end-users.

**Comment by jricher** [comment]
The AS and RS are related in this model. You are proposing a drastically different trust model, and not one that the working group has expressed interest in pursuing.

**Comment by Denisthemalice** [comment]
The trust relationships between the AS and the RS are currently undefined in the draft, as well as the trust relationships with the other components of the model.  See the issue "Trust relationships"" #214.

In this issue 214, I have proposed text for a new section to address the trust relationships. I would guess that you will have some objections with some sentences. It is the time to discuss about them. I invite you to reply and to propose your amendments if any.
At this time, you have not yet participated to the debate around the issue #214.

By experience, it is premature to design a protocol if the trust relationships are left undefined.  It is also impossible to correctly address privacy properties after the design of the protocols unless ending with a "Spy by design" architecture where the end-user consent and the user notices are ignored.

It is unfortunate, that the cart has been placed before the horse.

**Comment by agropper** [comment]
@Denisthemalice The trust model is indeed critical and a centerpiece of my privacy-driven comments in GNAP. Trust is coercive to the individual resource owner in the sense that a shrink-wrap license, or being forced to sign a HIPAA Privacy Practices Notice before getting care at a hospital is coercive to the sick and anxious person.

The GNAP trust model gives the RO as much power as possible in any particular use-case including, for example. US HIPAA and EU GDPR. The instrument of RO's power is control over the GNAP AS. That control can be self-sovereign if the AS is compiled from source and self-hosted or it can be based on a fiduciary relationship to the AS operator. GNAP need not be concerned with the fiduciary distinction and can reasonably assume that all ASs are self-sovereign to their owners.

Core aspects of RO power are delegation, automation, learning, and data minimization. Delegation enables the RO to hire experts as fiduciaries. Automation enables the RO to delegate to a semi-autonomous agent. Learning enables the RO and their agent adapt better and faster than the other actors in a use-case. Data minimization is one way that the RO and their agent protect their learning advantage and avoid surveillance by others, discrimination, or censorship. 

Surveillance is essential to the exercise of power. Ideally, the RO and their agent would have total information awareness (to use a Snowden-era term) but this is impractical in some use cases and may be illegal in law enforcement use cases such as No-fly lists.

A "Spy by design" architecture with the GNAP AS as the combined consent and surveillance actor is what I hope we're working toward. Giving other actors in the protocol the opportunity to contact the AS or the RO directly is desirable from this surveillance perspective as long as the power to avoid or minimize that leakage of information (the AS or RO service endpoint) such as through the introduction of pseudonyms and mediators under the RO's control.

**Comment by fimbault** [comment]
@agropper interesting input

**Comment by alanhkarp** [comment]
Tokens are not opaque to the client in SPKI, zcap-ld, Orie's implementation with VCs, or our Zebra Copy work.  That feature allows useful operations.  

- A client can delegate a sub-scope token to a process it runs.  For example, a renderer could be given only read permission even though the client has read/write.
- The RO may delegate tokens to its own AS that delegates them based on the user's policy. 
- Delegation can be done off-line.  The US Navy liked this feature for battlefield scenarios with uncertain connectivity.  For example, a token issued by an on-shore AS might need to be delegated while a ship at sea is radio silent.
- The client may want to verify that the token has only the specified scopes.

These benefits must be weighed against the advantages spelled out in this thread.  The biggest to my mind is the encapsulation that enables the issuing AS to change what's in a token without coordinating with anyone else.  Unfortunately, that precludes the RO from using its own AS for delegation of its tokens.

I think some benefits mentioned on this thread are less of an issue.  Privacy matters with non-opaque tokens can be handled by encrypting specific fields or by tokenizing those fields as done with credit card transactions.  Issuing tokens to one-off private keys preserves responsibility tracking while preventing unintended correlations.  

As an aside, the procedure specified in the Horton paper is more complicated than needed here.  Horton is doing responsibility tracking solely with object references in a programming language.  The necessary information can be encoded in a privacy preserving way directly into the tokens we're discussing.

**Comment by jricher** [comment]
@alanhkarp I completely disagree with this model. I think those benefits can be reached with the existing GNAP model, though in slightly different ways. Take the offline delegation for example -- why would the AS for the ship be on-shore? The ship would have its own AS close to the RS's. That ship's AS can take, as input, signals from on-short systems when available and make local decisions when not available. 

Additionally, much of the benefits of what's being described as a "token" in the above is not a good match for the "access token" in the current model. This is why I keep saying that things like zcap and VC's fit into two potential places:

 - as inputs to the AS during its delegation process; this is where I see the most application to what's proposed here
 - as token formats for the RS (not client) to consume; this is where the confusion comes in, as now it's tempting to push more expectations onto the client where they do not belong

Fundamentally, the client instance is the entity being delegated to, nothing else. The AS manages the delegation, and it does so on behalf of the RS when an RS exists (remember, we have identity use cases where there is no RS).

**Comment by agropper** [comment]
I'm not understanding the disagreement between @alanhkarp and @jricher and have not seen this Horton paper but I'm starting to see where @jricher and I could reach consensus. 

GNAP could design around each major actor having their own GNAP AS in the sense that a zero-trust architecture draws a trust boundary around a resource and its policy enforcement point. So, for example, 
- the RS (processor role) would have a GNAP AS it specifies and trusts, call it AS-RS,
- the RO (subject role) would have a GNAP AS it specifies and trusts, call it AS-RO,
- the RQ (delegate role) might have an AS of their own too but that seems far outside the GNAP protocol concerns.

As a result, when:
- the RO is interacting with the RS, they might actually be dealing with the AS-RS
- the RQ is interacting with the RS, they might be dealing with the AS-RS
- the RQ is interacting with the RO, they are dealing with the AS-RO
- the RS is interacting with the RO because they are presented with a token, they are dealing with the AS-RO
- the RS or RQ "need" to deal with the live RO rather than an agent they might have a way to bypass the AS-RO

Notice that I've avoided using the word "client" and that there could be many different kinds of identifiers, capabilities,  and tokens compatible with this model.

**Comment by alanhkarp** [comment]
Adrian Gropper ***@***.***> wrote:

> I'm not understanding the disagreement between @alanhkarp
> <https://github.com/alanhkarp> and @jricher <https://github.com/jricher>
> and have not seen this Horton paper but I'm starting to see where @jricher
> <https://github.com/jricher> and I could reach consensus.
>
I don't see it.

Justin wants the tokens to be opaque so that the AS-RS (the AS that the RS
trusts) is the only entity that needs to interpret them.  For example, if
the access token is a UUID, the AS-RS will maintain a table telling what
permissions it grants.  Later on, the AS-RS could decide to use a random
number instead of a UUID, and nobody else would have to change anything.
That's a huge advantage.

A downside is that all delegations have to go through AS-RS.  That works
but makes some useful capability patterns impossible or awkward.
Delegation is not possible if the delegator, such as AS-RO, an AS the RO
trusts, can't connect to AS-RS.  Also, delegations that you want to be
frequent and fast, such as a client delegating a sub-scoped access token to
a process it starts, will be impractically slow.

It looks like an either/or situation.  Either we give up some useful
capability patterns, or we have to deal with standard formats for the
access tokens and all the complication that entails.

--------------
Alan Karp

**Comment by agropper** [comment]
I guess I'm willing to give @jricher and @alanhkarp the benefit of the doubt. Trust, but verify.

The RS gets to decide what token formats it will support and whether it is willing to let each RO specify a different AS or forces the RO to pick from a list of "approved" ASs. 

A relationship between an RS and the AS-RS can be proprietary or GNAP-standard. As far as the RO and AS-RO, are concerned, as long as their relationship to RS and RQ are GNAP standard-based, how the RS decides to protect itself is their business. 

I would much prefer for every RS to honor my choice of AS-RO and let me decide on AS-first or RS-first interactions with the RQs. Therefore, the reason I see potential consensus with @jricher is because it might make it more likely that RS will adopt GNAP in the first place and honor my choice of AS-RO either directly or via their AS-RO.

**Comment by fimbault** [comment]
@agropper thxs for the clarification, if I summarize in my own words, you'd like to be able to optionally choose which AS is to be used. Is that correct?

Another use case for delegation through capabilities as @alanhkarp describes is for microservices/faas. One service might trigger many other scoped services, and JWTs are not well suited for the task (even if that's the de-facto solution today).

**Comment by agropper** [comment]
@fimbault Yes. But the choice of RO having a choice is still up to the RS. 

I definitely do not want to involve the **_typical_** RS in discovery of any kind with respect to the RO. That means the RS should be encouraged to use standards and publish openly (not behind any AS-RS, if they have one) everything they can about the services they provide, their policies, the structure of scopes they support, pricing, token types, hints, etc... All completely generic as far as specific ROs are concerned. 

The only thing a client can discover at an RS is an AS, if they bring some kind of locally unique identifier. Hence I support just the AS part of #187 but I prefer AS-first interactions in general so the typical client is just bringing the capability to the RS as their first encounter.

If the RO wants to disclose a DID, hints for successful requests, or anything else, the RO should find an RS that is explicitly designed as a directory service and may have nothing to do with the RS that is protecting the subject resource. The directory service could offer mediation services the way we expect of a dating service, for example.

Put another way that @Denisthemalice and @jricher might agree with, I would call an RS that has their own AS-RS **a directory or data broker.** GNAP should encourage directories and data brokers to use the protocol just as we encourage RS in a more tightly bound relationship to respect the RO choice of AS and any discovery mechanisms the RO might also choose. The ability for the RO to use a pseudonym or involve a mediator (like tor) when they approach a RS is, after all, one of the few things the RO can control and the RS can't.

**Comment by Denisthemalice** [comment]
The answer to the implicit question raised in the title of this thread is not YES or NO.

Some clients may configured to care about the end-user privacy while some other clients may be configured as "don't' care".

- Clients configured to care about the end-user privacy will refuse to forward opaque tokens since they want to verify the privileges that have been included into the access token. If the verification is fine, then they will forward the access token.
- Clients configured as "don't' care" will forward the access tokens whatever they are.

This approach is backwards compatible with OAuth 2.0 but offers new possibilities.

**Comment by fimbault** [comment]
I think all of these conversations make @jricher's proposal about the rewriting of section 4 and the idea of an extension document related to the RS a welcome contribution.

**Comment by Denisthemalice** [comment]
Before any re-writing, I would like to know whether **_jricher_** can agree with my previous message.

**Comment by jricher** [comment]
I think your question is invalid and misleading. It is based on the false premise that exposing information in the access token to the client is a design for privacy and that by not doing that you "don't care about privacy". This is inflammatory and accusatory language and is not helpful for the discussion.

**Comment by Denisthemalice** [comment]
The same two options would be phrased this way:

Some clients may configured to inspect the content of the access token and if the inspection is successful to forward it to the RS
while some other clients may be configured to directly forward the access token to the RS.

**Comment by alanhkarp** [comment]
Sorry, I neglected to send this one yesterday.  I just saw it in my Drafts
folder.  Watch for my comment on issue #145 for a followup.

Justin Richer ***@***.***> wrote:

> @alanhkarp <https://github.com/alanhkarp> I completely disagree with this
> model. I think those benefits can be reached with the existing GNAP model,
> though in slightly different ways. Take the offline delegation for example
> -- why would the AS for the ship be on-shore? The ship would have its own
> AS close to the RS's. That ship's AS can take, as input, signals from
> on-short systems when available and make local decisions when not available.
>
The ship case came directly from the Navy.  Their procedures start by
setting a policy for the ship before it leaves shore.  Once at sea, the
policy needs to be adapted based on local circumstances.  The ship has its
own AS for that purpose, but it's permissions come from the on-shore AS.
 Your answer in this case can be that those two AS's are in cahoots and can
understand the same opaque access tokens.

> Additionally, much of the benefits of what's being described as a "token"
> in the above is not a good match for the "access token" in the current
> model.
>
Sorry.  I intended "token" to be "access token" everywhere I used it except
when referring to tokenization.

> This is why I keep saying that things like zcap and VC's fit into two
> potential places:
>
>    - as inputs to the AS during its delegation process; this is where I
>    see the most application to what's proposed here
>
> As claims for VCs and in token exchange for zcaps, I presume.

>
>    - as token formats for the RS (not client) to consume; this is where
>    the confusion comes in, as now it's tempting to push more expectations onto
>    the client where they do not belong
>
> Fundamentally, the client instance is the entity being delegated to,
> nothing else. The AS manages the delegation, and it does so on behalf of
> the RS when an RS exists (remember, we have identity use cases where there
> is no RS).
>
Precluding off-line delegation might be the proper trade-off, but we should
understand what we're giving up.  Clients delegating small bits of their
permissions is key to enforcing least privilege at fine granularity.  Watch
https://www.youtube.com/watch?v=kdAHJTly6rI if you have the patience.  It's
fair to say that's out of scope for GNAP, but let's be explicit about the
decision.

--------------
Alan Karp

**Comment by alanhkarp** [comment]
@jricher is right.  Opaque tokens is the way to go.  Let me explain.

A group of us capability folks has been meeting every Friday morning for over 20 years.  One of the participants calls it Quaker computer science; we just sit around until someone pipes up with a topic.  It's a lot of fun, and I learn something almost every week.  This week I asked the opaque/non-opaque question.  The result was enlightening (and somewhat surprising to me.)  I also asked this question on a mailing list that reaches a broader community of capability folks.  I haven't heard from anyone, so I guess nobody on the list has a strong opinion.

The main issue I've had with opaque tokens is the inability to do offline delegation.  I asked the group how important that was, expecting to hear a number of important use cases.  The bottom line conclusion of the group was that offline delegation is a "nice to have" but not essential feature.  One person noted that might not have been the case 20 or even 10 years ago when continuous connectivity was less common.

One use case was that you were on safari when you wanted to delegate access to a companion about to return to civilization.  Another involved a client on Jupiter and an AS on Earth.  We all agreed that neither of those was compelling.  More importantly, we were able to find workarounds for almost all the realistic cases.  For example, most cases requiring low latency, such as a client delegating to a process it's starting, can be handled by doing the token exchange ahead of time.  In other cases, it may be possible for the client to proxy requests from the process and only forward those that it authorizes.  Opaque tokens won't be able to handle ad hoc, low latency sub-scope delegations.  Is that important?

One case does not have a workaround (yet?), partial communication, as in the safari use case.  A client gets an access token but cannot communicate with the RS or AS, perhaps due to a firewall.  The client would like to access the RS via a second client that it trusts only with a subset of its permissions.  Is this case common?  How important is it?

There was one issue that is better with non-opaque certificates as capabilities, e.g., zcap-ld, non-repudiation.  With opaque tokens, Alice will do a token exchange with the AS and give the new access token to Bob.  If Bob does something wrong, he can always claim it was Alice who used the capability.  With certificates, Alice will ask Bob for a public key, which Bob can create on the spot.  She will then create the delegation to that key.  There is no way for Alice to use that capability, so Bob can't blame her for actions he takes with it.  (Yes, I know.  More mechanism is needed.)  Is there a way in GNAP for Alice to tell the AS to deliver the new access token directly to Bob?  Is there a way for GNAP to deliver to Alice the access token encrypted with a public key provided by Bob?

We also did some brainstorming on how to do offline delegation with opaque tokens, which I've listed here for your amusement.  None of these ides has been given much thought.

- Ship with the access token mobile code that can create a new, potentially sub-scoped, token.
- Delegate by wrapping the access token with info on what subset of the token's scopes are being authorized.
- Create a grant that the delegatee can present to the AS to get the new access token.

**Comment by jricher** [comment]
To me, the Alice to Bob delegation makes more sense if we think of Bob being the new end user and his client instance getting its own token, but in the context of something that Alice has allowed. So Alice never sends Bob an access token at all, but instead could be given some other artifact that Bob can present to the AS as part of his own process of getting an access token. Bob never sees Alice's token, Alice never sees Bob's token, and there's no need for either. They could be managed completely separately, or Alice could be given control over revoking Bob's token because the rights of that token are based partially on her user delegation.

What does this artifact look like? I think that's where things like ZCAPs and VC's and the like really come into play. The AS can have any number of rules for who gets to access things -- could be an attribute, could be ability to present a known key, etc. The value here is that the AS remains in control over the policy of which tokens get issued, and it allows better security posture at the AS/RS. It could be as simple as Alice identifying "when Bob logs in with his google account bob@bob.com", which is more what we see today, but we aren't stuck with that limited model.

To me, it makes the most sense for GNAP's core protocol to address and specify how the client can facilitate these kinds of interactions with different parties. I think we've got all the hooks in there for something like this: the AS can ) return something that Alice can give to someone as a re-delegation artifact (in an undefined extension that defines the format and carrier, and Bob's client can present that artifact back at the AS as part of its request. We can even talk about the AS trying to reach out to Alice to confirm the request when Bob shows up, without Bob having to pretend that he's Alice in the process.

**Comment by alanhkarp** [comment]
Justin Richer ***@***.***> wrote:

> To me, the Alice to Bob delegation makes more sense if we think of Bob
> being the new end user and his client instance getting its own token, but
> in the context of something that Alice has allowed.
>
I agree, as long as we don't expect AS-RS to ever have heard of Bob before
issuing a token to him.

> So Alice never sends Bob an access token at all, but instead could be
> given some other artifact that Bob can present to the AS as part of his own
> process of getting an access token. Bob never sees Alice's token, Alice
> never sees Bob's token, and there's no need for either. They could be
> managed completely separately, or Alice could be given control over
> revoking Bob's token because the rights of that token are based partially
> on her user delegation.
>
I like it.  That's exactly what I did in my Client Utility work (the
project that led to HP's E-speak product.).  Alice gave Bob a nonce and
told AS-RS, "Whoever presents this nonce gets this subset of the
permissions in this access token."

In many cases it is important that Alice be able to revoke Bob's token and
all delegations derived from it.  That means AS-RS must keep track of the
delegations for two reasons.  One is to know all the downstream
delegations; another is to know if the revocation request is legitimate.
Permission to revoke can even be one of the scopes in an access token.

> What does this artifact look like? I think that's where things like ZCAPs
> and VC's and the like really come into play. The AS can have any number of
> rules for who gets to access things -- could be an attribute, could be
> ability to present a known key, etc.
>
 The artifact can be a certificate of some sort, but then you're back to
the kind of standards problems you run into with non-opaque tokens.  The
scheme we used for an HP internal system was opaque tokens, where all
permission information was held by the AS-RS.  Delegation involved
submitting your token with a list of scopes you were delegating.  Since the
scopes are part of the resource API, you have to understand them anyway.

> The value here is that the AS remains in control over the policy of which
> tokens get issued, and it allows better security posture at the AS/RS. It
> could be as simple as Alice identifying "when Bob logs in with his google
> account ***@***.***", which is more what we see today, but we aren't
> stuck with that limited model.
>
The policy expression language should be separate from the delegation
language.  The delegation language can be as simple as a list of scopes.
The policy language is typically more complex, as in many UMA examples.
Indeed, the policy language can be a differentiator for AS's bidding to be
chosed by an RS.

> To me, it makes the most sense for GNAP's core protocol to address and
> specify how the client can facilitate these kinds of interactions with
> different parties. I think we've got all the hooks in there for something
> like this: the AS can ) return something that Alice can give to someone as
> a re-delegation artifact (in an undefined extension that defines the format
> and carrier, and Bob's client can present that artifact back at the AS as
> part of its request.
>
I like it.

> We can even talk about the AS trying to reach out to Alice to confirm the
> request when Bob shows up, without Bob having to pretend that he's Alice in
> the process.
>
 That sounds a lot like OAuth 2.

--------------
Alan Karp

**Comment by jricher** [comment]
Here I'll use the term "AS-RS" to refer to what I've been calling the "AS" in the rest of the conversation, because I'm talking specifically about something that is tied directly to the RS.

> I agree, as long as we don't expect AS-RS to ever have heard of Bob before issuing a token to him.

I would even push it further to say that AS-RS might not have ever heard of Alice, either. 

>In many cases it is important that Alice be able to revoke Bob's token and all delegations derived from it. That means AS-RS must keep track of the delegations for two reasons. One is to know all the downstream delegations; another is to know if the revocation request is legitimate. Permission to revoke can even be one of the scopes in an access token.

Revocation permission, and particularly chained revocation, is something that the AS-RS can offer. 

> The artifact can be a certificate of some sort, but then you're back to the kind of standards problems you run into with non-opaque tokens. 

That's only true if you consider the artifact to be a "token". I am drawing a bright line between what the AS-RS consumes and what the AS-RS produces. The first half can be non-opaque to the client, and could even be generated by the client or an agent or some other component in a known fashion. It could also be the kind of opaque nonce that you talk about, too, something the client instance has to trade in. Both models work as different kinds of inputs to the AS-RS.

However, what the AS-RS produces is an opaque access token, which is what this issue is about. Specifically, it's opaque to the client instance -- it may or may not be opaque to the RS, but since the AS and RS have a trust relationship that's for them to figure out. These two domains are orthogonal to each other.

> The policy expression language should be separate from the delegation language. The delegation language can be as simple as a list of scopes. The policy language is typically more complex, as in many UMA examples. Indeed, the policy language can be a differentiator for AS's bidding to be chosed by an RS.

Remember, Bob's client will be asking for a set of rights (in the `access_token.access` field) alongside presenting whatever artifact(s) Alice gave him. The delegation will be in terms of this access field, which replaces the `scope` concept from OAuth 2. The policy might be more externalized, the language of the artifact.

> We can even talk about the AS trying to reach out to Alice to confirm the request when Bob shows up, without Bob having to pretend that he's Alice in the process.
> That sounds a lot like OAuth 2.

Not at all. OAuth 2 is about Alice's software client not pretending it's Alice the person. OAuth 2 has no idea who Bob is, at all. 

OAuth 2 is fundamentally about delegation from a user to a piece of software. GNAP expands that to delegating from a user or possibly set of users to a piece of software, while potentially facilitating more advanced use cases like delegating to a different user's piece of software using those mechanisms.

**Comment by alanhkarp** [comment]
I agree.

--------------
Alan Karp

**Comment by Denisthemalice** [comment]
The discussion i:s drifting away from its tittle which is" Opaque / Non opaque access tokens". 

My position is rather simple : 
there exist cases where the client is willing to take a look at the content of the access token and 
other cases where the client is not interested to take a look at the content of the access token. 

All these cases should be supported by GNAP.
 
For the negotiation of the access token formats, clients should be able to know which access token formats are supported 
by an AS and by a RS.

**Comment by fimbault** [comment]
I agree with Denis that both cases could occur. I believe that it's ok for the core spec to only deal with opaque tokens, and handle additional possibilities (in particular token format) via extensions.

**Comment by Denisthemalice** [comment]
**@ fimbault** I believe that it's ok for the core spec to deal **_both_** with opaque tokens and non-opaque tokens,

Token format negotiation is necessary as soon as there exist no prior relationships between a RS and an AS 
(i.e. for system components on the Internet). 

In terms of trust relationships; 
- the RS **_must_** trust  the AS for "performing some actions", while 
- the AS **_may_** trust some ROs that have been dubbed by a RS.

**Comment by jricher** [comment]
@Denisthemalice token opacity is always a matter of perspective. Opacity to the client is what's under discussion here, and I strongly believe tokens should remain opaque to the client for the purposes of the core protocol. Tokens need not be opaque to the RS, and in the case where the RS and AS don't have a prior relationship it should be both discoverable and negotiable. This is under the purview of the current section 10 which the editors have discussed pulling out into a second draft for better focus.

**Comment by alanhkarp** [comment]
Denisthemalice ***@***.***> wrote:

> *@ fimbault* I believe that it's ok for the core spec to deal *both* with
> opaque tokens and non-opaque tokens,
>
> Token format negotiation is necessary as soon as there exist no prior
> relationships between a RS and an AS
> (i.e. for system components on the Internet).
>
I don't understand what you're saying here.  Clearly, the RS must trust the
AS that validates tokens the RS receives, which I've been calling AS-RS.
The RS doesn't have to trust any other AS, such as the one RO trusts to
enforce RO's policy, which I've been calling  AS-RO.  All that matters is
that the access token is a proper delegation.  In neither case is token
format negotiation required for opaque tokens.

> In terms of trust relationships;
>
>    - the RS *must* trust the AS for "performing some actions", while
>    - the AS *may* trust some ROs that have been dubbed by a RS.
>
> Why must AS-RS trust some ROs?  The AS will give the RO one or more access
tokens based on what the RO presents to AS-RS, which may be a credential or
money.  I don't believe there's a need for any trust between RO and AS-RS
beyond that.

--------------
Alan Karp

>

**Comment by fimbault** [comment]
The issue is that everyone brings its own interpretation of the boundary between RS/AS. Hence a proposal from the editors to keep a simple model in place, which is known to work.

**Comment by agropper** [comment]
@fimbault I don't understand your statement and what you mean by a "simple model". 
- Are you objecting to @alanhkarp terminology of AS-RS and AS-RO as being overly complex? If so, how do we deal with RO policies AND minimizing the ability for AS-RS to learn about requests that may not result in any token or an attenuated access token?

**Comment by fimbault** [comment]
"simple" is subjective of course, but I meant close in spirit to the current draft. 
I think more advanced flows are something to consider, but probably as an extension.

**Comment by agropper** [comment]
What part of what I'm asking about (in terms of power and privacy of the RO) would be considered an advanced flow?

**Comment by fimbault** [comment]
That's just a personal view here: the AS-RO part would fall in an advanced flow.

**Comment by Denisthemalice** [comment]
**_@jricher_** You wrote:

> I strongly believe tokens should remain opaque to the client for the purposes of the core protocol. 

However, you don't provide any technical argument. Do you want to take privacy considerations 
out of the scope of the core protocol ?

Personally, I prefer to consider them from the very beginning. I dislike "privacy after design".

This does not mean that all clients **_should_** or **_shall_** inspect the content of the access tokens, 
but that some clients **_may_** be willing to do so.

**Comment by jricher** [comment]
@Denisthemalice Many people, including me, have given you many technical arguments. 

I also have stated, and repeat here, that having the client software able to inspect the token is an anti-privacy practice, since in many instances (which again have been explained to you many times) the client is the untrusted component. Your proposal is, as you would say, "spy by design" by the client software. Even allowing that is problematic. There are even use cases where exposing all information to an individual RS is a privacy violation, which is why some systems opt to use differential introspection (each RS gets a different, limited answer for what the token's for). 

This protocol and this group are not doing "privacy after design" as you assert. Privacy is a core component of what's being built, and privacy considerations are central to this protocol. Just because the WG disagrees with your proposed solution does not mean the working group is rejecting privacy as a principle.

**Comment by jricher** [comment]
@agropper the way I see it, the simple case is when all the policies live internally to the AS. It should be possible for the AS to externalize these and have that be an input into the process, but we shouldn't expect that of all deployments. Fundamentally, what you're asking for doesn't affect the format of the access token, but does affect how the AS operates. I think there are some nuances there and it's going to take some additional discussion, but we can look at it in the context of a rewrite of the current Section 4 that's been proposed by the editors. I'll remember to tag you on that PR once we have something available.

**Comment by agropper** [comment]
@jricher I agree that "the simple case is when all the policies live internally to the AS." I also agree that the format of the access token should not be affected.

In OAuth and OpenID Connect the RO is forced to share their policies with an AS chosen by the RS, if they want to delegate to a policy-driven AS. The RO is offered an opportunity to participate live in the authorization process but they are not afforded the power of delegation to an agent of their choice. 

I look forward to your PR.

**Comment by Denisthemalice** [comment]
**@jricher** You wrote: 

> Privacy is a core component of what's being built, and privacy considerations are central to this protocol.

I am glad to read such a sentence. However, at this moment the Security Considerations section is only 3 lines thick. 

You also wrote:

> the client is the untrusted component.

Such sentence does not describe a trust relationship nor an absence of trust relationship. I will recall, once again, that a trust relationship, when it exists, MUST be described using three elements A, B and C : 

**_The entity A is trusting the entity B for some kind of action C._** 

An end-user trusts its client to behave as intended, i.e. according to the protocols and to display correctly the results of the exchanges when appropriate. If an end-user does not trust his laptop, he should not use it. 

As long as the trust relationships will be missing in the draft, we will not be able to progress. See issue #214

**Comment by fimbault** [comment]
@Denisthemalice we know the security and privacy considerations are not written yet, there are issues open for that. We can expect the trust relationship of the core GNAP to remain relatively close to what OAuth2 proposes. In particular we'd be handling the the client as an untrusted component (which makes sense because many attacks come from malicious clients). 

As soon as time allows, there are plans for extensions which might implement some of the ideas mentioned, like those described by Alan/Adrian/you etc. (all of those will need proper review when time comes, and test implementations would help clarify). I would expect to find there support for token format, AS-RO, etc.

**Comment by Denisthemalice** [comment]
**_@fimbault_**. You expect the trust relationship of the core GNAP to remain relatively close to what OAuth2 proposes.

This sentence is quite interesting. 

In OAuth 2.0 Authorization Framework (RFC 6750) the word _**trust**_ is not even present in the core of the document. 

In The OAuth 2.1 Authorization Framework (draft-ietf-oauth-v2-1-00) the situation is not much better even if there are a few occurrences of the word **_trust_** in the core of the document. It is impossible to know which component of the model is trusting which other component and for which reason.

See the issue #214 where I propose text to address this issue. 
If you have amendments to propose for this text, then it is the right time to do it.

**Comment by fimbault** [comment]
@Denisthemalice I will have a look at issue 214 in more details

**Comment by agropper** [comment]
@fimbault I respect the intent of the editors and the difficulty of your task but dealing with core privacy issues through "extensions that might implement some of the ideas mentioned" seems unwarranted. It risks creating a schism between W3C and IETF as both groups try to represent key privacy rights of the individual.

For example, over on the W3C / DIF standards, we discuss Confidential Storage in the context of encrypted data vaults as RSs with the express purpose of minimizing the information that the RS can glean from their role as a service provider. The data is encrypted, indexes are encrypted, and access control under consideration is based on capabilities. 

In that same workgroup we also consider cases where the RS is processing non-encrypted data. This introduces complications around discovery which then must also be managed through access control. 

A privacy and community perspective in both W3C and IETF suggests that the RO should have the power to delegate as the core, not by "extensions". If we take that perspective, there's hope that access control across processors (resource servers) of all kinds, ranging from encrypted data vaults to directories and data brokers will be consistent and supported by self-sovereign and fiduciary controllers (authorization servers).

**Comment by fimbault** [comment]
@agropper Thanks for the feedback, I hear that. The intent is not to create a schism, quite the contrary we intend to allow compatibility between various well known standards (IETF, W3C, OpenID, DIF, etc.). I can speak for editors when say that we do intend to support privacy, as much as possible. We value your proposals, as well as those of others who participated in this thread and on the mailing list. 

There are however other considerations to take into account (my personal view here):
- the paint is not completely fresh, so it's good to take a bit of time to review proposals. The AS-RO delegation scheme seems neat (and I do like capabilities) but as we've seen there are other competing solutions that people bring to the table (including for instance, my very own proposal for hints, which comes with the same scrutiny by the WG, which I think is great). There are various ways each component can be designed, and we also need to be careful in what we include. Probably we should also take the time to prototype a little bit more.
- the core spec should focus on simplicity, while still enabling future enhancements. If we see we can indeed support more "human centric flows" in the core itself to cover many business cases against a reasonable complexity, I'll be happy to support that. 
- based on the feedback from the WG, there are already big changes ahead for sections 4 and 10 that we'd need to get right before we can reasonably get into additional parts.

**Comment by agropper** [comment]
@fimbault Again, I appreciate the clarity of your language and your honorable intent and look forward to the changes the editors will propose.

FWIW, I have spent about a decade implementing authorization around the "other considerations" you mention. I have dealt with OAuth and OpenID Connect at scale using the largest healthcare examples (the Epic EHR and US Federal Medicare database). I've been a core contributor to UMA2 and helped found the HEART WG to profile OAuth APIs in the healthcare context. 

If GNAP does not support "human centric flows" at our core we will be driving people ever deeper into the Solid Pod https://solidproject.org model where authorization is tied to processing as in AS-RS and we expect people to own and lease Pods from their service providers. In healthcare, this model looks like Apple Health where I am forced to copy my data (from Epic or Medicare) to my iPhone and then I am subject to hidden and capricious app store policies for ever getting that data out via API. Meanwhile, in the healthcare field we have UDAP https://www.udap.org busy building extensions / profiles to OAuth.

My point is, we know what AS-RS looks like.

**Comment by fimbault** [comment]
I don't disagree ;-)

A recap that this issue is currently pending close since last week and about to close.

This thread has been uber long, and we now tend to get away from the initial discussion. I'd suggest we get those items (e.g. AS-RO) in their own thread, so that we make sure we keep them to the top of our list.

On the specific opaque/non-opaque item: 
- to recap the reasons from the editors, [last week](https://github.com/ietf-wg-gnap/gnap-core-protocol/issues/145#issuecomment-801209794).
- I don't think there's any substantial change since then, on this specific discussion point
 
Thanks for the many interesting inputs (many of which are related to a larger scope) that we'll definitely take into account. The core spec will handle opaque tokens, and we'll keep the conversation open for token format descriptions as extensions (here the boundary seems fairly clear).

**Comment by alanhkarp** [comment]
Denisthemalice ***@***.***> wrote:

> *The entity A is trusting the entity B for some kind of action C.*
>
>
I like your definition of trust, but sometimes it's easier to understand if
we turn things around.  In what way is entity A vulnerable to entity B?  In
this case, B might perform C to A's expectations, but B might also do
action D, which A did not expect.  A privacy violation is an example of D.

--------------
Alan Karp

>

**Comment by Denisthemalice** [comment]
**@alanhkarp** 

I am glad you like the definition, which is a shortcut of the ISO definition.  :-)

Trust relationships are necessary to define secure protocols. However, alone there are insufficient to address privacy issues.

For example, if an end-user **_A_** CANNOT trust **two RSs** for NOT attempting to link their user accounts by using the "sub_id" claim that is present in a JWT access token, then the end-user A should ask to an AS to use a "sub_id" claim identifier that contains a value that is specific to every AS-RS pair.

Considering such negative trust relationships allows to define mechanisms and protocols able to take care of the end-user privacy.

When we will have identified the complete set of  both **_trust relationships_** and **_non-trust relationships_**, 
then we will be able to build the protocols ... and then the exercise will be rather easy.

**@fimbault** 

You wrote:

> The core spec will handle opaque tokens, and we'll keep the conversation open for token format descriptions as extensions

I fear you are proposing to place the topic about **_non opaque tokens "under the carpet"_**.

This would mean preventing the client (and thus the end-user) to be able to verify that the AS behaves as intended.

If you really intend to do so, you should first fill-in the Privacy Considerations section and ask to the WG if your text addition to this (nearly) empty section is adequate or  not.

**Comment by fimbault** [comment]
No, I'm not putting the topic under the carpet. The choice is clear for the core spec, and I clearly support having the ability to define token formats as an extension.

**Comment by jricher** [comment]
Consensus is that access tokens will remain opaque to the client instance and AS-RS connections will be discussed in new document, #224
